include/deal.II/numerics/vector_tools.h

```diff
@@ -2516,6 +2516,82 @@ namespace VectorTools
+    const NormType &                                         norm,
+    const Function<spacedim, double> *                       weight   = nullptr,
+    const double                                             exponent = 2.);
+
+  /**
+   * Same as above for hp.
+   */
+  template <int dim, class InVector, class OutVector, int spacedim>
+  void
+  integrate_difference(
+    const hp::MappingCollection<dim, spacedim> &             mapping,
+    const hp::DoFHandler<dim, spacedim> &                    dof,
+    const InVector &                                         fe_function,
+    const Function<spacedim, typename InVector::value_type> &exact_solution,
+    OutVector &                                              difference,
+    const hp::QCollection<dim> &                             q,
+    const NormType &                                         norm,
+    const Function<spacedim, double> *                       weight   = nullptr,
+    const double                                             exponent = 2.);
+
+  /**
+   * Call the integrate_difference() function, see above, with
+   * <tt>mapping=MappingQGeneric@<dim@>(1)</tt>.
+   */
+  template <int dim, class InVector, class OutVector, int spacedim>
+  void
+  integrate_difference(
+    const hp::DoFHandler<dim, spacedim> &                    dof,
+    const InVector &                                         fe_function,
+    const Function<spacedim, typename InVector::value_type> &exact_solution,
+    OutVector &                                              difference,
+    const hp::QCollection<dim> &                             q,
+    const NormType &                                         norm,
+    const Function<spacedim, double> *                       weight   = nullptr,
+    const double                                             exponent = 2.);
+
+  /**
+   * Compute the cellwise error of the finite element solution.  Integrate the
+   * difference between a reference function which is given as a continuous
+   * function object, and a finite element function. The result of this
+   * function is the vector @p difference that contains one value per active
+   * cell $K$ of the triangulation. Each of the values of this vector $d$
+   * equals
+   * @f{align*}{
+   * d_K = \| u-u_h \|_X
+   * @f}
+   * where $X$ denotes the norm chosen and $u$ represents the exact solution.
+   *
+   * @deprecated Use integrate_difference(const Mapping<dim, spacedim> &, const DoFHandler<dim, spacedim> &, const InVector &, const Function<spacedim, typename InVector::value_type> &, OutVector &, const Quadrature<dim> &, const NormType &, const Function<spacedim, double> *, const double) instead.
+   */
+  template <int dim, class InVector, class OutVector, int spacedim>
+  DEAL_II_DEPRECATED typename std::enable_if<
+    !std::is_same<typename InVector::value_type, double>::value>::type
   integrate_difference(const Mapping<dim, spacedim> &    mapping,
```

Comment:
I used `std::enable_if` for the deprecated functions.  The deprecated function would be used for example if `InVector` is `float` and `Function` is `double`. If `InVector` is `double` and `Function` is `double` then the new function would be used.

If I don't use `std::enable_if` I run into conflicts.